clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -488,7 +491,11 @@ private void close(KafkaChannel channel) {
         } catch (IOException e) {
             log.error("Exception closing connection to node {}:", channel.id(), e);
         }
-        this.stagedReceives.remove(channel);
+        Deque<NetworkReceive> deque = this.stagedReceives.remove(channel);
+        if (deque != null) {
+            addToCompletedReceives(channel, deque, true);
```

Comment:
With producer acks=0, when the producer sends messages and closes the connection without waiting for responses, selector can process both received data and close in the same poll().